### PR TITLE
Jax bug fixes for the dot product attention

### PR DESCRIPTION
**Pull request by zlsh80826**
This PR includes several bug fixes related to the fused and the unfused attention, and also not enabling the fused attention by default.

1. Fixed an issue where fusing the scale and softmax operations with a bias resulted in incorrect computation flow. The fix ensures that the computation flow becomes Softmax(scale * attn_weights + bias). e594bb625b40deb9a5206e3ac4431832fe187967
2. There is a bug that not clearing `S` tensor leads to incorrect answer on cuDNN kernel. WAR by clearing `S` tensor when `causal_masking` + `no_bias` and add the  relevant unit tests. ac67e914b6e3bbd6a6cdbbbdf073026a1e86e9f2
3. Enhance the handling of `None` sharding tensor. 49e73d51f2dae2aeee056c709f8efbdc39f51d73
4. Based on internal discussions, it has been decided to disable fused attention by default for JAX in the upcoming release (0.9). This PR introduces a new flag, `NVTE_USE_FUSED_ATTN`, which enables fused attention for internal convergence tests (t5x/PAXML). The flag will be removed once the convergence tests are successfully completed. ac67e914b6e3bbd6a6cdbbbdf073026a1e86e9f2
5. Add `thread_local` to protect the global static plan cache for thread safety.  8a0d11a8dc0057ee3cf8a392ad27f2a7a4cea6ae

### CODE REVIEW

**Comment by mingxu1067** [comment]
LGTM

**Comment by nouiz** [comment]
Can any of those fix affect t5x converges?

**Comment by zlsh80826** [comment]
> Can any of those fix affect t5x converges?

Yes, the first item is a bug fix for t5x converge, the t5x will not converge (non-fmha, scaledSoftmax) without that fix.
The item 2,3,5 is for PAXML/GPT, not for t5x.

**Comment by nouiz** [comment]
It would be great to have tests for the bugfixes.

**Comment by nouiz** [comment]
Otherwise the small issue, it looks good to me.

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -261,20 +278,22 @@ def grad_func(fused_attn_max_512_func, *args, **kwargs):
...
-            atol=3e-5)
-
-        # dbeta padded part
-        np.testing.assert_allclose(
-            jnp.asarray(primitive_dbeta[:, :, self.valid_len:, self.valid_len:], np.float32),
-            jnp.asarray(reference_dbeta[:, :, self.valid_len:, self.valid_len:], np.float32))
-
-        assert jnp.allclose(primitive_dbeta[:, :, self.valid_len:, self.valid_len:],
-                            jnp.zeros_like(primitive_dbeta[:, :, self.valid_len:, self.valid_len:]))
+        if self.attn_bias_type != AttnBiasType.NO_BIAS:
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply tests/jax/test_fused_attn.py:281]: Because if inputs don't include bias, there's nothing to check for dbias

[the remark this answers is not in the document]
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:281]: I think I should rename `*_dbeta` to `*_dbias` for the naming alignment. I will change it later.

[the remark this answers is not in the document]
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:281]: Renamed in https://github.com/NVIDIA/TransformerEngine/pull/236/commits/57d97ad3c64e7da2c7b154fc9fab8b4ba5c461b7

**Comment by zlsh80826** [comment]
> It would be great to have tests for the bugfixes. I mean, only the second commit have a test.

Yes, I agree. I added a unit test for item 1. (4040ae36260a990a399bca85c1d453eedee971c1). For item 3,5, I will look into whether those can be convered by the multi-card unit tests in the future (other PRs).

File: tests/jax/test_fused_attn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nouiz) [reply tests/jax/test_fused_attn.py:281]: Thanks.

File: transformer_engine/common/fused_attn/fused_attn_fp16_bf16_max_seqlen_512.cu [hunk 2]
```diff
@@ -668,7 +667,8 @@ void fused_attn_max_512_fwd_impl(int64_t b, int64_t h, int64_t s_q, int64_t s_kv
             createScale(b, h, s_q, s_kv, d, layout, tensorType, ops);
 
             // if bias, we need to memset the S buffer to correctly computate dbias
-            auto zero_s = (bias_type != NVTE_Bias_Type::NVTE_NO_BIAS);
+            auto zero_s = (bias_type != NVTE_Bias_Type::NVTE_NO_BIAS) ||
+                          (mask_type == NVTE_Mask_Type::NVTE_CAUSAL_MASK);
```
Comment (timmoon10) [inline transformer_engine/common/fused_attn/fused_attn_fp16_bf16_max_seqlen_512.cu:671]: I'd like to understand this bugfix a bit more. The cuDNN bug only shows up with `causal_mask` and not the other mask modes?

**Review by timmoon10 — APPROVED** [review]
LGTM

**Comment by timmoon10** [comment]
Can you explain why you needed to make the plan caches `thread_local`? Could you accomplish thread-safety by protecting access with mutexes, or is there a more fundamental reason why threads shouldn't share the same resources? I wonder if the JIT kernel infrastructure needs to make a similar change.

File: transformer_engine/common/fused_attn/fused_attn_fp16_bf16_max_seqlen_512.cu (hunk 2 above)
Comment (jeng1220) [reply transformer_engine/common/fused_attn/fused_attn_fp16_bf16_max_seqlen_512.cu:671]: Yes, see `nvbugs 4122719` for details.

**Comment by jeng1220** [comment]
> Can you explain why you needed to make the plan caches `thread_local`? Could you accomplish thread-safety by protecting access with mutexes, or is there a more fundamental reason why threads shouldn't share the same resources? I wonder if the JIT kernel infrastructure needs to make a similar change.

It is according to cuDNN design. The cuDNN handle shouldn't not be shared by multiple host threads, and the plans associating with the handle shouldn't either.

We did hit the error if one thread executes a plan generated by another thread before.

https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#thread-safety

**Review by cyanguwa — APPROVED** [review]
Looks good to me. Thanks!

**Review by mingxu1067 — APPROVED** [review]

File: transformer_engine/jax/flax/transformer.py [hunk 3]
```diff
@@ -360,12 +370,13 @@ def kv_init(key, shape, dtype):
         q_seqlen = inputs_q.shape[0] if self.transpose_batch_sequence else inputs_q.shape[1]
         kv_seqlen = inputs_kv.shape[0] if self.transpose_batch_sequence else inputs_kv.shape[1]
         fused_attn_supported_seqlen = [128, 256, 384, 512]
+        enable_fused_attn = int(os.getenv("NVTE_USE_FUSED_ATTN", "0"))
```
[the remark this answers is not in the document]
Comment (zlsh80826) [reply transformer_engine/jax/flax/transformer.py]: Changed in 74eb2b38e091752b4ca21e1f8336469d05547d09

**Comment by zlsh80826** [comment]
Hello @timmoon10, @cyanguwa,

I think this PR is ready, could you help merge this for 0.9 release? Thanks!
